SpatialConvolutionLocal.lua

```diff
@@ -0,0 +1,166 @@
+local SpatialConvolutionLocal, parent = torch.class('nn.SpatialConvolutionLocal', 'nn.Module')
+
+function SpatialConvolutionLocal:__init(nInputPlane, nOutputPlane, iW, iH ,kW, kH, dW, dH, padW, padH)
+   parent.__init(self)
+
+   dW = dW or 1
+   dH = dH or 1
+
+   self.nInputPlane = nInputPlane
+   self.nOutputPlane = nOutputPlane
+   self.kW = kW
+   self.kH = kH
+   self.iW = iW
+   self.iH = iH
+
+   self.dW = dW
+   self.dH = dH
+   self.padW = padW or 0
+   self.padH = padH or self.padW
+   self.oW = math.ceil((self.padW * 2 + iW - self.kW) / self.dW + 1)
+   self.oH = math.ceil((self.padH * 2 + iH - self.kH) / self.dH + 1)
+   assert(1 <= self.oW and 1 <= self.oH, 'illegal configuration: output width or height less than 1')
+
+   self.weight = torch.Tensor(self.oH, self.oW, nOutputPlane, nInputPlane, kH, kW)
+   self.bias = torch.Tensor(nOutputPlane, self.oH, self.oW)
+   self.gradWeight = torch.Tensor():resizeAs(self.weight)
+   self.gradBias = torch.Tensor():resizeAs(self.bias)
+
+   self.finput = torch.Tensor()
+   self.fgradInput = torch.Tensor()
+
+   self:reset()
+end
+
+function SpatialConvolutionLocal:reset(stdv)
+   if stdv then
+      stdv = stdv * math.sqrt(3)
+   else
+      stdv = 1/math.sqrt(self.kW*self.kH*self.nInputPlane)
+   end
+   if nn.oldSeed then
+      self.weight:apply(function()
+         return torch.uniform(-stdv, stdv)
+      end)
+      self.bias:apply(function()
+         return torch.uniform(-stdv, stdv)
+      end)
+   else
+      self.weight:uniform(-stdv, stdv)
+      self.bias:uniform(-stdv, stdv)
+   end
+end
+
+local function makeContiguous(self, input, gradOutput)
+   if not input:isContiguous() then
+      self._input = self._input or input.new()
+      self._input:resizeAs(input):copy(input)
+      input = self._input
+   end
+   if gradOutput then
+      if not gradOutput:isContiguous() then
+	 self._gradOutput = self._gradOutput or gradOutput.new()
+	 self._gradOutput:resizeAs(gradOutput):copy(gradOutput)
+	 gradOutput = self._gradOutput
+      end
+   end
+   return input, gradOutput
+end
+
+local function viewWeight(self)
+   self.weight = self.weight:view(self.oH * self.oW, self.nOutputPlane, self.nInputPlane * self.kH * self.kW)
+   if self.gradWeight and self.gradWeight:dim() > 0 then 
+      self.gradWeight = self.gradWeight:view(self.oH * self.oW, self.nOutputPlane, self.nInputPlane * self.kH * self.kW)
+   end
+end
+
+local function unviewWeight(self)
+   self.weight = self.weight:view(self.oH, self.oW, self.nOutputPlane, self.nInputPlane, self.kH, self.kW)
+   if self.gradWeight and self.gradWeight:dim() > 0 then 
+      self.gradWeight = self.gradWeight:view(self.oH, self.oW, self.nOutputPlane, self.nInputPlane, self.kH, self.kW)
+   end
+end
+
+local function checkInputSize(self, input)
+   if input:nDimension() == 3 then
+      if not input:isSize(torch.LongStorage{self.nInputPlane, self.iH, self.iW}) then
```

Comment:
you are allocating a new tensor at every forward pass. I think it would be better to just do
```lua
input:size(1) == self.nInputPlane and input:size(2) == self.iH and input:size(3) == self.iW
```
Same thing for the `checkOutputSize` function